File: photonlib-java-examples/src/main/java/org/photonlib/examples/simposeest/robot/Robot.java
/*
 * Copyright (C) Photon Vision.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

package org.photonlib.examples.simposeest.robot;

import edu.wpi.first.wpilibj.TimedRobot;
import edu.wpi.first.wpilibj.geometry.Pose2d;
import edu.wpi.first.wpilibj.kinematics.ChassisSpeeds;
import org.photonlib.examples.simposeest.sim.DrivetrainSim;

public class Robot extends TimedRobot {

    AutoController autoCtrl = new AutoController();
    Drivetrain dt = new Drivetrain();
    OperatorInterface opInf = new OperatorInterface();

    DrivetrainSim dtSim = new DrivetrainSim();

    PoseTelemetry pt = new PoseTelemetry();

    @Override
    public void robotInit() {
        // Flush NetworkTables every loop. This ensures that robot pose and other values
        // are sent during every iteration.
        setNetworkTablesFlushEnabled(true);
    }

    @Override
    public void autonomousInit() {
        resetOdometery();
        autoCtrl.startPath();
    }

    @Override
    public void autonomousPeriodic() {
        ChassisSpeeds speeds = autoCtrl.getCurMotorCmds(dt.getCtrlsPoseEstimate());
        dt.drive(speeds.vxMetersPerSecond, speeds.omegaRadiansPerSecond);
        pt.setDesiredPose(autoCtrl.getCurPose2d());
    }

    @Override
    public void teleopPeriodic() {
        dt.drive(opInf.getFwdRevSpdCmd(), opInf.getRotateSpdCmd());
    }

    @Override
    public void robotPeriodic() {
        pt.setEstimatedPose(dt.getCtrlsPoseEstimate());
        pt.update();
    }

    @Override
    public void disabledPeriodic() {
        dt.drive(0, 0);
    }

    @Override
    public void simulationPeriodic() {
        if (opInf.getSimKickCmd()) {
            dtSim.applyKick();
        }
        dtSim.update();
        pt.setActualPose(dtSim.getCurPose());
    }

    private void resetOdometery() {
        Pose2d startPose = autoCtrl.getInitialPose();
        dtSim.resetPose(startPose);
        dt.resetOdometry(startPose);
    }
}

File: photonlib-cpp-examples/src/main/cpp/examples/getinrange/cpp/Robot.cpp
/*
 * Copyright (C) Photon Vision.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

#include "Robot.h"

#include <photonlib/PhotonUtils.h>

void Robot::TeleopPeriodic() {
  double forwardSpeed;
  double rotationSpeed = xboxController.GetLeftX();

  if (xboxController.GetAButton()) {
    // Vision-alignment mode
    // Query the latest result from PhotonVision
    photonlib::PhotonPipelineResult result = camera.GetLatestResult();

    if (result.HasTargets()) {
      // First calculate range
      units::meter_t range = photonlib::PhotonUtils::CalculateDistanceToTarget(
          CAMERA_HEIGHT, TARGET_HEIGHT, CAMERA_PITCH,
          units::degree_t{result.GetBestTarget().GetPitch()});

      // Use this range as the measurement we give to the PID controller.
      forwardSpeed =
          -controller.Calculate(range.value(), GOAL_RANGE_METERS.value());
    } else {
      // If we have no targets, stay still.
      forwardSpeed = 0;
    }
  } else {
    // Manual Driver Mode
    forwardSpeed = -xboxController.GetRightY();
  }

  // Use our forward/turn speeds to control the drivetrain
  drive.ArcadeDrive(forwardSpeed, rotationSpeed);
}

#ifndef RUNNING_FRC_TESTS
int main() { return frc::StartRobot<Robot>(); }
#endif
